### DISCUSSION TITLE
guidelines for intellij plugin that depends on org.intellij.scala plugin

### DISCUSSION THREAD

**Discussion by techLeadAtLogicOverCode** [JetBrains/sbt-idea-plugin#105] [Q&A] [answered]
There is a "fluent sbt project template" intellij plugin https://github.com/techLeadAtLogicOverCode/FluentSbtTemplatePlugin 
that generates a new sbt project with a pre defined sbt plugin and fluent style build definition in build.sbt.   

Now, this intellij plugin uses the "new project sbt wizard" that is a part of "intellij scala" plugin. To use that wizard,
"fluent sbt project template" intellij plugin has to depend on "intellij scala" plugin.   

As, it is easy to understand in audio visual mode, So, created a 3 minutes youtube video, to explain "fluent sbt project template" intellij plugin and it's dependency on "intellij scala" plugin.   

Please watch the video in high-definition(hd) mode :-

https://youtu.be/CGC3Yq390xk

---

<b>Functional requirement :-</b>

As it is clear from video, that the simplified "fluent sbt project template" intellij plugin depends on "intellij scala" plugin.

As and when a new release of "intellij scala" plugin will be available. A new release of "fluent sbt project template" intellij plugin also needs to be created.   
For the same, following files needs to be changed in "fluent sbt project template" intellij plugin :-   

<i>

- source code
- project/build.properties
- project/plugins.sbt
- META-INF/plugin.xml
- build.sbt
  
</i>

Below are the questions related to above files that are present in "fluent sbt project template" intellij plugin.

---

<b>source code</b>

Source code changes in "fluent sbt project template" intellij plugin can be done by following source code of "intellij scala" plugin.

---

<b>project/plugins.sbt</b>

"intellij scala" plugin project/plugins.sbt uses 3.11.0 version of sbt-idea-plugin

`addSbtPlugin("org.jetbrains" % "sbt-idea-plugin" % "3.11.0")`

"fluent sbt project template" intellij plugin uses 3.12.2 version of sbt-idea-plugin

`addSbtPlugin("org.jetbrains" % "sbt-idea-plugin" % "3.12.2")`

<u>Que.</u>

<span style="color:red">
<i>Kindly confirm, if it is ok, or the sbt-idea-plugin version should be same ?</i>   
</span>

---

<b>project/build.properties</b>

`sbt.version=1.3.13`

Referring whatever version is available in "intellij scala" plugin project/build.properties.

<u>Que.</u>

<span style="color:red">
<i>Kindly confirm, if it is mandatory for the sbt versions to be same if "fluent sbt project template" intellij plugin depends on "intellij scala" plugin ?</i>   
</span>

---

<b>META-INF/plugin.xml</b>

`<idea-version since-build="211.5538" until-build="212.*"/>` 

idea version in "fluent sbt project template" intellij plugin is referred from "intellij scala" plugin below link :

https://github.com/JetBrains/intellij-scala/blob/840ec4ae3b6d52662454f2fe15077ed6d634a8d3/scala/scala-impl/resources/META-INF/plugin.xml#L39

<u>Que.</u>

<span style="color:red">
<i>Not sure, if idea version required exactly to be same. Please confirm ?</i>   
</span>

Also, what is the relation between :-   

<p align="center">
plugin.xml
</p>

`<idea-version since-build="211.5538" until-build="212.*"/>` 

<p align="center">
and
</p>

<p align="center">
build.sbt
</p>

`intellijBuild in ThisBuild := "211.6693.14"`

in "intellij scala" plugin ?

---

<b>build.sbt</b>

<u>Que 1.</u>

<span style="color:red"><i>
Please suggest, which scalaVersion to use in "fluent sbt project template" intellij plugin, as there are different scalaVersion in build.sbt file of "intellij scala" plugin ?
</i>

https://github.com/JetBrains/intellij-scala/blob/idea211.x/build.sbt

</span>

<u>Que 2.</u>

<span style="color:red"><i>
Is it fine to set intellijDownloadSources to false ?
Please suggest best practice if "fluent sbt project template" intellij plugin depends on 
"intellij scala" plugin ? 
</i>  
`intellijDownloadSources := false`
</span>

<u>Que 3.</u>

`bundleScalaLibrary in ThisBuild := false`

<span style="color:red"><i>
Why it is required to set bundleScalaLibrary key to false as suggested in below link ?
</i></span>
https://gitter.im/JetBrains/intellij-scala?at=5f18098ca4b25c73c1e19240

<u>Que 4.</u>

<span style="color:red"><i>
Which java version should I use for the development of "fluent sbt project template" intellij plugin ?
</i></span>

<u>Que 5.</u>

<span style="color:red"><i>
What are the other configuration, that needs to be referred if "fluent sbt project template" intellij plugin depends on "intellij scala" plugin ? 
</i></span>

**Comment by mutcianm** [comment, marked as answer]
## > Kindly confirm, if it is ok, or the sbt-idea-plugin version should be same?
No, not necessarily. Feel free to use the latest available release of `sbt-idea-plugin`.
You're essentially using IJ Scala plugin as a **binary dependency** for compiling your own new IJ plugin, thereby the Scala plugin's build-time code will not affect the way you build your own plugin.


## > Kindly confirm, if it is mandatory for the sbt versions to be same if "fluent sbt project template" intellij plugin depends on "intellij scala" plugin?

No, it's not mandatory.

## >Not sure, if idea version required exactly to be same. Please confirm?

No, they're not required exactly to be same. Your plugin's `idea-version` should only be used to enforce version compatibility of **your own** plugin. IJ platform will take care of updating the dependent plugins for you.

## >Also, what is the relation between ... `intellijBuild` in the build.sbt and the `since-build` in plugin.xml?

- `intellijBuild` only tells sbt which IJ build to download and compile your sources against.
- `since-build` tells  the IJ plugin manager that your plugin can be loaded in IJ products version `since-build` and newer

So this means that:
- `since-build` cannot be newer than `intellijBuild` or your plugin simply won't be loaded when you run in via sbt's `runIDE`
- `since-build` can be used to mitigate compatibility issues arising from unexpected IJ platform API breakages after updating to a newer `intellijBuild` by releasing a version of your plugin which fixes the issues and updating the `since-build` so the update will only be installable on the newer IJ builds


## >Please suggest, which scalaVersion to use in "fluent sbt project template" intellij plugin, as there are different scalaVersion in build.sbt file of "intellij scala" plugin?

The `scalaVersion` used by the Scala plugin is defined in the [`dependencies.scala`](https://github.com/JetBrains/intellij-scala/blob/idea211.x/project/dependencies.scala#L4). For Scala plugin major version `2021.1` Scala "2.13.2" is used. Although this may change at some point in a newer major version. Unfortunately at this point it must be tracked by the plugin authors manually, see https://github.com/JetBrains/sbt-idea-plugin/issues/17

## >Is it fine to set `intellijDownloadSources` to false ?
If sbt fails do download the sources than you may disable it ofc. However you'll loose the ability to inspect IJ platform sources then.
I'd suggest picking such an `intellijBuild` that has sources available. The latest **stable** release version should be fine.  You can find the versions [here](https://www.jetbrains.com/idea/download/other.html).
As for the sources of the Scala plugin, it is unfortunately impossible to automatically attach them to your project at the moment(see https://github.com/JetBrains/sbt-idea-plugin/issues/55). You can manually download the sources archive from the GH of the Scala plugin and attach them from IDEA manually when prompted.


## >Why it is required to set bundleScalaLibrary key to false as suggested in below link ?

There was a bug in the `sbt-idea-plugin` that caused scala-library to be bundled even if there was a dependency on the Scala plugin for IJ(which already provides an instance of scala-library). This setting key should now be initialized with a correct value automatically, depending on whether your plugin depends on some other plugin that already provides scala-library or not. The reason for this key's existence is explained [here](https://github.com/JetBrains/sbt-idea-plugin#bundlescalalibrary-in-thisbuild--settingkeyboolean).

In general you should not need to modify this key, but if in this case scala-library is getting bundled in your plugin anyway, please report an issue.

## > Which java version should I use for the development of "fluent sbt project template" intellij plugin ?
JDK 11

## >What are the other configuration, that needs to be referred if "fluent sbt project template" intellij plugin depends on "intellij scala" plugin?

Looks like you've got all covered already. I'd suggest checking out the ["Known Issues"](https://github.com/JetBrains/sbt-idea-plugin#known-issues-and-limitations) section though
